src/Language/PureScript/Linter.hs

```diff
@@ -129,3 +132,136 @@ lint (Module _ _ mn ds _) = censor (addHint (ErrorInModule mn)) $ mapM_ lintDecl
     | name `S.member` s = (s, errorMessage' ss (mkError name))
     | otherwise = (S.insert name s, mempty)
 
+
+
+lintUnused :: forall m. (MonadWriter MultipleErrors m) => Module -> m ()
+lintUnused (Module modSS _ mn modDecls exports) =
+  censor (addHint (ErrorInModule mn)) $ do
+    topVars <- traverse lintDeclaration modDecls
+    let allVars = S.unions topVars
+    case exports of
+      Nothing ->
+        pure ()
+      Just exports' -> do
+        let n (ValueRef _ ident) = Just ident
+            n _ = Nothing
+            exportIds = S.fromList $ mapMaybe n exports'
+            expectedUsedDecls = S.fromList (mapMaybe getDeclIdent modDecls) `S.difference` exportIds
+            newErrs = mconcat $ map (errorMessage' modSS . UnusedDeclaration) $ S.toList $ expectedUsedDecls `S.difference` allVars
+        tell newErrs
+        pure ()
+  where
+  getDeclIdent :: Declaration -> Maybe Ident
+  getDeclIdent (ValueDeclaration vd) = Just (valdeclIdent vd)
+  getDeclIdent (ExternDeclaration _ ident _) = Just ident
+  getDeclIdent (TypeInstanceDeclaration _ _ _ ident _ _ _ _) = Just ident
+  getDeclIdent BindingGroupDeclaration{} = internalError "lint: binding groups should not be desugared yet."
+  getDeclIdent _ = Nothing
```

Comment:
This is `getIdentName <=< declName`.

Reply:
This is copied from earlier in the same file actually, oops!